warehouse/query-core/src/main/java/datawave/query/common/grouping/GroupingUtil.java

```diff
@@ -83,25 +84,24 @@ public GroupingInfo getGroupingInfo(Map.Entry<Key,Document> entry, Set<String> g
                                 int count = ((BigDecimal) countTypeAttribute.getType().getDelegate()).intValue();
                                 countKeyMap.put(countKey, count);
                             });
-            
-            Multimap<String,String> fieldToFieldWithContextMap = getFieldToFieldWithGroupingContextMap(entry.getValue(), expandedGroupFieldsList, fieldMap,
-                            groupFieldsSet, reverseModelMapping);
+            Set<String> groupingContextSet = new HashSet<>();
+            TreeMultimap<String,String> fieldToFieldWithContextMap = (TreeMultimap<String, String>) getFieldToFieldWithGroupingContextMap(entry.getValue(), expandedGroupFieldsList, fieldMap,
+                            groupFieldsSet, groupingContextSet, reverseModelMapping);
             log.trace("got a new fieldToFieldWithContextMap: {}", fieldToFieldWithContextMap);
             int longest = longestValueList(fieldToFieldWithContextMap);
-            for (int i = 0; i < longest; i++) {
+            for (String currentGroupingContext : groupingContextSet) {
                 Collection<GroupingUtil.GroupingTypeAttribute<?>> fieldCollection = new HashSet<>();
-                String currentGroupingContext = "";
                 for (String fieldListItem : expandedGroupFieldsList) {
                     log.trace("fieldListItem: {}", fieldListItem);
-                    Collection<String> gtNames = fieldToFieldWithContextMap.get(fieldListItem);
+                    NavigableSet<String> gtNames = fieldToFieldWithContextMap.get(fieldListItem);
                     if (gtNames == null || gtNames.isEmpty()) {
```

Comment:
gtNames apparently can't be null.  TreeMultiMap returns an empty set for missing keys.